### PR TITLE
More smartly determine `MAX_JOBS`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
The default `MAX_JOBS` should make the build as fast as possible, while still run out of memory. I tested locally with gcc 13.1, with 64 jobs in parallel, the highest memory usage is 181GB, so roughly 2.8GB/task. To be safer, we should consider capping the max jobs to `available mem / 3`.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -298,9 +323,17 @@ def cmake(build_dir: str = "", install_prefix: str = "./nvfuser"):
...
+    available_mem = get_available_memory_gb()
+    if available_mem > 0:
+        max_jobs_mem = int(available_mem / mem_gb_per_task)
+        max_jobs = min(max_jobs, max_jobs_mem)
+
     if not CMAKE_ONLY:
         # build binary
-        max_jobs = os.getenv("MAX_JOBS", str(multiprocessing.cpu_count()))
+        max_jobs = os.getenv("MAX_JOBS", str(max_jobs))
+        print(f"Using {max_jobs} jobs for compilation")
```
Comment (jjsjann123) [inline setup.py:336]: I thought we have a logging system. Looks like I was just hallucinating.

File: setup.py [hunk 2]
```diff
@@ -251,6 +251,31 @@ def version_tag():
     return version
 
 
+def get_available_memory_gb():
```
Comment (jjsjann123) [inline setup.py:254]: nitpick, maybe this one should go to `tools` instead?

**Review by jjsjann123 — APPROVED** [review]
I love stamping!

**Comment by jjsjann123** [comment]
a few nitpicks. feel free to merge as-is.

File: setup.py (hunk 1 above)
Comment (zasdfgbnm) [reply setup.py:336]: We do, except that we defined a `logger` and just ignored it and keep using `print` in later codebase :-D

Comment (zasdfgbnm) [reply setup.py:336]: I am trying to use this `logger` but the message no longer show up in the console. I am pretty sure there is a way to do so, but I will just use `print` for now.

File: setup.py (hunk 2 above)
Comment (zasdfgbnm) [reply setup.py]: moved

**Comment by zasdfgbnm** [comment]
!build

**Pull request merged**